nomad/job_endpoint_hook_implicit_identities.go

```diff
@@ -111,31 +110,27 @@ func (h jobImplicitIdentitiesHook) handleConsulTasks(t *structs.Task, consul *st
 //  1. The task has a Vault block.
 //  2. The server is configured with `vault.use_identity = true` and a
 //     `vault.default_identity` is provided.
-//
-// If the task already has an identity named `vault` it sets the identity name
-// to the expected value.
 func (h jobImplicitIdentitiesHook) handleVault(t *structs.Task) {
 	if !h.srv.config.UseVaultIdentity() || t.Vault == nil {
 		return
 	}
 
 	// Use the Vault identity specified in the task.
-	for _, wid := range t.Identities {
-		if wid.Name == vaultIdentityName {
-			return
-		}
+	vaultWID := t.GetIdentity(t.Vault.IdentityName())
+	if vaultWID != nil {
+		return
 	}
 
 	// If the task doesn't specify an identity for Vault, fallback to the
 	// default identity defined in the server configuration.
-	vaultWID := h.srv.config.VaultDefaultIdentity()
+	vaultWID = h.srv.config.VaultDefaultIdentity()
 	if vaultWID == nil {
 		// If no identity is found skip inject the implicit identity and
 		// fallback to the legacy flow.
 		return
 	}
 
 	// Set the expected identity name and inject it into the task.
-	vaultWID.Name = vaultIdentityName
+	vaultWID.Name = t.Vault.IdentityName()
```

Comment:
A minor thing, but when we call this we have to allocate a new buffer for the string each time, and we call it twice in the same function. Might be nice if we called `vaultIdentiyName := t.Vault.IdentityName()` somewhere near the top of the function.

Reply:
Fixed in https://github.com/hashicorp/nomad/pull/18591/commits/d8c142c34a7fbecd09b64151ab4a6f80b5971bd3. This method is still not used much, but I will keep this in mind in future PRs.